paddle/operators/fc_op.cc

```diff
@@ -0,0 +1,173 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/framework/op_registry.h"
+#include "paddle/operators/net_op.h"
+
+namespace paddle {
+namespace operators {
+
+class FCOp : public NetOp {
+ public:
+  FCOp(const std::string &type, const framework::VariableNameMap &inputs,
+       const framework::VariableNameMap &outputs,
+       const framework::AttributeMap &attrs)
+      : NetOp(type, inputs, outputs, attrs) {
+    auto x = Inputs("X");
+    auto w = Inputs("W");
+    auto mul_out = Outputs("MulOut");
+    PADDLE_ENFORCE_EQ(
+        x.size(), w.size(),
+        "The size of inputs X(%d) should be the same as that of weights W(%d).",
+        x.size(), w.size());
+    PADDLE_ENFORCE_EQ(mul_out.size(), x.size(),
+                      "The size of intermediate mul_out(%d) should be the same "
+                      "as that of inputs X(%d).",
+                      mul_out.size(), x.size());
+
+    size_t n = x.size();
+    PADDLE_ENFORCE_GE(n, static_cast<size_t>(1),
+                      "The size of inputs X(%d) should be no less than 1.", n);
+
+    auto x_num_col_dims = Attr<std::vector<int>>("xNumColDims");
+    PADDLE_ENFORCE_EQ(x_num_col_dims.size(), n,
+                      "The size of attribute xNumColDims(%d) should be the "
+                      "same as that of inputs X(%d).",
+                      x_num_col_dims.size(), n);
+
+    // mul_out[i] = X[i] * W[i]
+    for (size_t i = 0; i < n; i++) {
+      framework::AttributeMap mul_attr;
+      mul_attr["x_num_col_dims"] = static_cast<int>(x_num_col_dims[i]);
+      mul_attr["y_num_col_dims"] = static_cast<int>(1);
+      AppendOp(
+          framework::OpRegistry::CreateOp("mul", {{"X", {x[i]}}, {"Y", {w[i]}}},
+                                          {{"Out", {mul_out[i]}}}, mul_attr));
+    }
+
+    // sum_out = X[0] * W[0] + ... + X[n-1] * W[n-1]
+    if (n > 1) {
+      AppendOp(framework::OpRegistry::CreateOp(
+          "sum", {{"X", {mul_out}}}, {{"Out", {Output("SumOut")}}}, {}));
+    } else {
+      AppendOp(framework::OpRegistry::CreateOp(
+          "identity", {{"X", {mul_out[0]}}}, {{"Y", {Output("SumOut")}}}, {}));
+    }
+
+    // add_out = sum_out + b
+    auto b = Input("B");
+    std::string add_out = "SumOut";
+    if (b != framework::kEmptyVarName) {
+      add_out = "AddOut";
+      AppendOp(framework::OpRegistry::CreateOp(
+          "rowwise_add", {{"X", {Output("SumOut")}}, {"b", {Input("B")}}},
+          {{"Out", {Output(add_out)}}}, {}));
+    } else {
+      if (Output("AddOut") != framework::kEmptyVarName) {
+        this->Rename(Output("AddOut"), framework::kEmptyVarName);
+      }
+    }
+
+    auto activation = Attr<std::string>("activation");
+    AppendOp(framework::OpRegistry::CreateOp(
+        activation, {{"X", {Output(add_out)}}}, {{"Y", {Output("Y")}}}, {}));
```

Comment:
Please reference https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/name_convention.md

It seems that our `Output` prefer to `Out`.

Reply:
I notice that. But I prefer to use `Y` :disappointed: Whatever, I will rename it.